File: JS/question3/Connect4.ts
export class Connect4 {
  private board: number[][];
  private currentPlayer: number;

  constructor() {
    this.board = Array.from({ length: 6 }, () => new Array<number>(7).fill(0));
    this.currentPlayer = 1;
  }

  public play(column: number): string {
    if (column < 0 || column >= 7) return "Invalid column!";
    if (this.isGameFinished()) return "Game has finished!";

    for (let row = 5; row >= 0; row--) {
      if (this.board[row][column] === 0) {
        this.board[row][column] = this.currentPlayer;
        const result = `Player ${this.currentPlayer} has a turn`;

        if (this.isWinningMove(row, column)) {
          return `Player ${this.currentPlayer} wins!`;
        }

        this.currentPlayer = 3 - this.currentPlayer;
        return result;
      }
    }

    return "Column full!";
  }

  public reset(): void {
    this.board = Array.from({ length: 6 }, () => new Array(7).fill(0));
    this.currentPlayer = 1;
  }

  private isWinningMove(row: number, col: number): boolean {
    for (let i = -3; i <= 0; ++i) {
      // Check horizontally
      if (
        col + i >= 0 &&
        col + i + 3 < 7 &&
        this.board[row][col + i] === this.currentPlayer &&
        this.board[row][col + i + 1] === this.currentPlayer &&
        this.board[row][col + i + 2] === this.currentPlayer &&
        this.board[row][col + i + 3] === this.currentPlayer
      ) {
        return true;
      }

      // Check vertically
      if (
        row + i >= 0 &&
        row + i + 3 < 6 &&
        this.board[row + i][col] === this.currentPlayer &&
        this.board[row + i + 1][col] === this.currentPlayer &&
        this.board[row + i + 2][col] === this.currentPlayer &&
        this.board[row + i + 3][col] === this.currentPlayer
      ) {
        return true;
      }

      // Check diagonally from top-left to bottom-right
      if (
        row + i >= 0 &&
        row + i + 3 < 6 &&
        col + i >= 0 &&
        col + i + 3 < 7 &&
        this.board[row + i][col + i] === this.currentPlayer &&
        this.board[row + i + 1][col + i + 1] === this.currentPlayer &&
        this.board[row + i + 2][col + i + 2] === this.currentPlayer &&
        this.board[row + i + 3][col + i + 3] === this.currentPlayer
      ) {
        return true;
      }

      // Check diagonally from bottom-left to top-right
      if (
        row - i < 6 &&
        row - i - 3 >= 0 &&
        col + i >= 0 &&
        col + i + 3 < 7 &&
        this.board[row - i][col + i] === this.currentPlayer &&
        this.board[row - i - 1][col + i + 1] === this.currentPlayer &&
        this.board[row - i - 2][col + i + 2] === this.currentPlayer &&
        this.board[row - i - 3][col + i + 3] === this.currentPlayer
      ) {
        return true;
      }
    }

    return false;
  } 

  private isGameFinished(): boolean {
    for (let row = 0; row < 6; row++) {
      for (let col = 0; col < 7; col++) {
        if (this.board[row][col] !== 0 && this.isWinningMove(row, col)) {
          return true;
        }
      }
    }
    return false;
  }
}

File: C#/I3SolutionsTask3/Connect4.cs
namespace I3SolutionsTask3;

public static class Connect4
{

    private static int[,] board = new int[6, 7];
    private static int currentPlayer = 1;

    public static string Play(int column)
    {
        if (column < 0 || column >= 7)
            return "Invalid column!";


        if (IsGameFinished())
            return "Game has finished!";

        for (int row = 5; row >= 0; row--)
        {
            if (board[row, column] == 0)
            {
                board[row, column] = currentPlayer;


                string result = $"Player {currentPlayer} has a turn";


                if (IsWinningMove(row, column))
                    return $"Player {currentPlayer} wins!";


                currentPlayer = 3 - currentPlayer;


                return result;
            }
        }

        // Column is full
        return "Column full!";
    }

    public static void Reset()
    {
        board = new int[6, 7];
        currentPlayer = 1;
    }

    private static bool IsWinningMove(int row, int col)
    {//Checking all directions

        // Check horizontally
        for (int i = -3; i <= 0; ++i)
        {
            if (col + i < 0 || col + i + 3 > 6) continue;
            if (board[row, col + i] == currentPlayer &&
                board[row, col + i + 1] == currentPlayer &&
                board[row, col + i + 2] == currentPlayer &&
                board[row, col + i + 3] == currentPlayer)
            {
                return true;
            }
        }

        // Check vertically
        for (int i = -3; i <= 0; ++i)
        {
            if (row + i < 0 || row + i + 3 > 5) continue;
            if (board[row + i, col] == currentPlayer &&
                board[row + i + 1, col] == currentPlayer &&
                board[row + i + 2, col] == currentPlayer &&
                board[row + i + 3, col] == currentPlayer)
            {
                return true;
            }
        }

        // Check diagonally from top-left to bottom-right
        for (int i = -3; i <= 0; ++i)
        {
            if (row + i < 0 || row + i + 3 > 5 || col + i < 0 || col + i + 3 > 6) continue;
            if (board[row + i, col + i] == currentPlayer &&
                board[row + i + 1, col + i + 1] == currentPlayer &&
                board[row + i + 2, col + i + 2] == currentPlayer &&
                board[row + i + 3, col + i + 3] == currentPlayer)
            {
                return true;
            }
        }

        // Check diagonally from bottom-left to top-right
        for (int i = -3; i <= 0; ++i)
        {
            if (row - i > 5 || row - i - 3 < 0 || col + i < 0 || col + i + 3 > 6) continue;
            if (board[row - i, col + i] == currentPlayer &&
                board[row - i - 1, col + i + 1] == currentPlayer &&
                board[row - i - 2, col + i + 2] == currentPlayer &&
                board[row - i - 3, col + i + 3] == currentPlayer)
            {
                return true;
            }
        }

        return false;
    }


    private static bool IsGameFinished()
    {
        // Check if any player has won
        for (int row = 0; row < 6; ++row)
        {
            for (int col = 0; col < 7; ++col)
            {
                if (board[row, col] != 0 && IsWinningMove(row, col))
                {
                    return true;
                }
            }
        }

        // Check if board is full
        for (int col = 0; col < 7; ++col)
        {
            if (board[0, col] == 0)
            {
                return false;
            }
        }

        return true;
    }
}